contracts/coordinator/src/state.rs

```diff
@@ -1,18 +1,86 @@
 use crate::error::ContractError;
 use cosmwasm_schema::cw_serde;
 use cosmwasm_std::{Addr, Order, Storage};
-use cw_storage_plus::{Index, IndexList, IndexedMap, Map, MultiIndex};
+use cw_storage_plus::{Index, IndexList, IndexedMap, MultiIndex, UniqueIndex};
 use router_api::ChainName;
 use std::collections::HashSet;
 
 pub const CONTRACT_NAME: &str = env!("CARGO_PKG_NAME");
 pub const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
 type ProverAddress = Addr;
+type VerifierAddress = Addr;
 
-pub const PROVER_PER_CHAIN: Map<ChainName, ProverAddress> = Map::new("prover_per_chain");
+struct ChainProverIndexes<'a> {
+    pub by_chain: UniqueIndex<'a, ChainName, ChainProverRecord, (ChainName, ProverAddress)>,
+    pub by_prover: UniqueIndex<'a, ProverAddress, ChainProverRecord, (ChainName, ProverAddress)>,
+}
+
+impl<'a> IndexList<ChainProverRecord> for ChainProverIndexes<'a> {
+    fn get_indexes(&self) -> Box<dyn Iterator<Item = &dyn Index<ChainProverRecord>> + '_> {
+        let v: Vec<&dyn Index<ChainProverRecord>> = vec![&self.by_chain, &self.by_prover];
+        Box::new(v.into_iter())
+    }
+}
+
+#[cw_serde]
+pub struct ChainProverRecord {
+    pub prover: ProverAddress,
+    pub chain: ChainName,
+}
+
+const CHAIN_PROVER_INDEXED_MAP: IndexedMap<
+    (ChainName, ProverAddress),
+    ChainProverRecord,
+    ChainProverIndexes,
+> = IndexedMap::new(
+    "chain_prover_map",
+    ChainProverIndexes {
+        by_chain: UniqueIndex::new(|d| d.chain.clone(), "chain_prover_map_by_chain"),
```

Comment:
why do we need this index? The underlying map uses chain as a primary key, so can't we just look up directly?